File: typescript/Behavioral/14-Command/Invoker/JobActionInvoker.ts
import { ICommand } from "../Command/ICommand";

export class JobActionInvoker {
  private readonly history: ICommand[] = [];
  private readonly queue: ICommand[] = [];

  enqueue(command: ICommand): void {
    this.queue.push(command);
  }

  executeNext(): void {
    if (this.queue.length === 0) {
      console.log("  Queue is empty.");
      return;
    }
    const command = this.queue.shift()!;
    command.execute();
    this.history.push(command);
  }

  executeAll(): void {
    while (this.queue.length > 0) this.executeNext();
  }

  undo(): void {
    if (this.history.length === 0) {
      console.log("  Nothing to undo.");
      return;
    }
    this.history.pop()!.undo();
  }
}

File: csharp/Behavioral/14-Command/Invoker/JobActionInvoker.cs
public class JobActionInvoker
{
    private readonly Stack<ICommand> _history = new();
    private readonly Queue<ICommand> _queue   = new();

    public void Enqueue(ICommand command) => _queue.Enqueue(command);

    public void ExecuteNext()
    {
        if (_queue.Count == 0) { Console.WriteLine("  Queue is empty."); return; }
        var command = _queue.Dequeue();
        command.Execute();
        _history.Push(command);
    }

    public void ExecuteAll()
    {
        while (_queue.Count > 0) ExecuteNext();
    }

    public void Undo()
    {
        if (_history.Count == 0) { Console.WriteLine("  Nothing to undo."); return; }
        _history.Pop().Undo();
    }
}
